plugins/performance-lab/includes/server-timing/class-perflab-server-timing.php

```diff
@@ -225,6 +225,25 @@ public function use_output_buffer(): bool {
 		return (bool) apply_filters( 'perflab_server_timing_use_output_buffer', $enabled );
 	}
 
+	/**
+	 * Adds hooks to send the Server-Timing header.
+	 *
+	 * When output buffering is enabled, buffer as early as possible so that any other plugins that also do output
+	 * buffering will be able to register Server-Timing metrics. The first output buffer callback to be registered
+	 * is the last one to be called, so by starting the Server-Timing output buffer as soon as possible we can be
+	 * assured that other plugins' output buffer callbacks will run before the Server-Timing one that sends the
+	 * Server-Timing header.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function add_hooks(): void {
+		if ( $this->use_output_buffer() ) {
+			add_action( 'template_redirect', array( $this, 'start_output_buffer' ), PHP_INT_MIN );
+		} else {
+			add_filter( 'template_include', array( $this, 'on_template_include' ), PHP_INT_MAX );
+		}
```

Comment:
I'm curious why we need to change the hook that this is attached to depending on if OB is being used? I also wonder why we just don't move the hook regardless of the OB status if there is a benefit, so that the timing will be consistent.

The other consideration here is that the `wp-before-template` metrics are calculated from the `tempate_include` hook as well ([reference](https://github.com/WordPress/performance/blob/trunk/plugins/performance-lab/includes/server-timing/defaults.php#L71-L80)) and I think we want these to happen on the same hook in order to accurately report the `wp-total` value.

Reply:
> I'm curious why we need to change the hook that this is attached to depending on if OB is being used? I also wonder why we just don't move the hook regardless of the OB status if there is a benefit, so that the timing will be consistent.

When output buffering is not enabled, the `Server-Timing` header should be sent as late as possible to give a chance for plugins to register their metrics. In this case, `template_include` filter with max priority is appropriate. 

On the other hand, when output buffer _is_ enabled then we still need to send the `Server-Timing` header at the very end, but the way we do so is different: we do so by starting the output buffer _earlier_ so that it wraps the entire page resulting in its callback running at the very end.

If we always sent the `Server-Timing` header at `template_redirect` (even when not doing output buffering) then this would potentially exclude plugins from being able to register their metrics since they may occur between `template_redirect` and `template_include`.

> The other consideration here is that the `wp-before-template` metrics are calculated from the `tempate_include` hook as well ([reference](https://github.com/WordPress/performance/blob/trunk/plugins/performance-lab/includes/server-timing/defaults.php#L71-L80)) and I think we want these to happen on the same hook in order to accurately report the `wp-total` value.

I'm not sure the change impacts this `wp-before-template` metrics calculation since the change here is only changing when output buffering starts, not when calculation happens.

Reply:
Ah, right, the metric still gets recorded at the same time, it's just when we start the buffer that changes. Makes sense.

The interesting side effect here is that this allows us to add Server Timing headers to things like `robots.txt`, feeds, etc. which weren't previously possible.